### DISCUSSION TITLE
R - Bagaimana menggunakan pivot_longer() kepada dataframe dengan columns nama berulang?

### DISCUSSION THREAD

**Discussion by Hertzspr** [teamalgoritma/community#67] [Q&A] [answered]
Saya punya dataframe seperti ini
![Capture](https://user-images.githubusercontent.com/95233522/164696566-5c98bab4-0c71-4145-a2df-0345aade7e25.JPG)

1. Bagaimana supaya 3 kolom day_of_week bergabung menjadi 1 kolom "best_day_of_week"?
2. Bagaimana supaya 3 kolom bermacam total bergabung menjadi 1 kolom "metrics"?
3. Kemudian value isian bermacam kolom total jadi kolom paling kanan dengan nama "total" 

[wideanswer.csv](https://github.com/teamalgoritma/community/files/8540023/wideanswer.csv)

**Comment by NabiilahArdini** [comment]
Halo Bapak Husada @Hertzspr ,

Berikut saya coba dengan beberapa tahapan:

```
library(dplyr)
library(stringr)
library(tidyr)

data <- read.csv("wideanswer.csv")

# filter data utuk ambil nama kolom
# dari data, ambil yang nama kolom datanya terdeteksi punya pattern tertentu

day <- data %>% select_if(str_detect(names(data), "day"))
stats <- data %>% select_if(str_detect(names(data), "total"))
```

Saya menggunakan skema piping (%>%) dari dplyr sehingga data sebelum piping akan digunakan untuk proses setelah piping. Disitu juga terdapat fungsi `select_if()` untuk menseleksi kolom sesuai kondisi. Kondisinya saya isi untuk bila terdeteksi nama kolom dari data terdapat pattern "day" dan "total". Fungsi `str_detect()` berasal dari package **sringr**. Dua data yang dihasilkan akan memudahkan saya untuk mengambil nama kolom di tahap pivot selanjutnya.

```
# 2 step pivot
pivot_longer(data, cols = names(day), values_to = "best_day_of_week") %>%
  pivot_longer(cols = names(stats), names_to = "metrics", values_to = "total") %>% 
  select(-name) # tidak diperlukan lagi
```
Tahapan di atas menghasilkan:

<img width="416" alt="image" src="https://user-images.githubusercontent.com/51941102/164704272-13d86102-1b8a-4924-9d41-21c8c6c65a8d.png">

Bapak boleh coba jalankan pertahapan piping agar bisa melihat proses kerja tiap code nya. 

Demikian yang dapat saya sampaikan. Sampai saat ini saya belum terpikir/ketemu cara yang lebih singkat lagi. 
Namun semoga cukup membantu Bapak.

Best regards,
Nabiilah

**Reply by Hertzspr** [reply]
Terima kasih.

**Reply by Hertzspr** [reply]
Waduh mbak, baru sadar, ini hasilnya Friday, Tuesday, dan Thursday malah jadi sama semua ya. 
Padahal maksudnya:
1. untuk total stream dengan value 229 terbaik jatuh di Friday, 
2. untuk total listener dengan value 47 terbaik jatuh di Tuesday, 
3. untuk total followers_diff dengan value 3, terbaik jatuh di Thursday.

**Reply by Hertzspr** [reply]
Atau bisa diambil baris 1,5,9 aja ya, kalau terlanjur begitu?

**Reply by NabiilahArdini** [reply]
Ini bisa dilakukan tapi sepertinya kurang reproducible. Saya kurang menyarankan. Kecuali jika kita bisa buatkan fungsi/rule pengambilan barisnya dengan sekuens tertentu.

**Comment by NabiilahArdini** [comment, marked as answer]
Halo Bapak Husada, 

Menanggapi reply Bapak di jawaban sebelumnya, sepertinya saya baru menangkap struktur data yang Bapak maksud. Bila seperti itu, maka pivot bisa dilakukan terpisah untuk day dan metrics, kemudian baru digabungkan menjadi 1 data. Namun untuk proses ini, Bapak haru memastikan bahwa urutan kolomnya saling bersebelahan agar ketika di pivot dan digabungkan datanya bisa match dengan tepat. Berikut codenya:

```
# pivot day
pivot_day <- pivot_longer(data, cols = names(day), values_to = "best_day_of_week") %>% 
  select(best_day_of_week)

# pivot stats
pivot_stats <- pivot_longer(data, cols = names(stats), names_to = "metrics", values_to = "total") %>% 
  select(metrics, total)

# join
bind_cols(pivot_stats, pivot_day)
```
<img width="345" alt="image" src="https://user-images.githubusercontent.com/51941102/165007079-c8ac4346-bb4c-49ea-8119-e217e53c3d6c.png">

**Reply by Hertzspr** [reply]
`cols = names(day)` dan `cols = names(day)` nya error mbak, saya ganti `cols = "day_of_week"` dan `cols = c("total_streams","total_listeners","total_followers_diff")` bisa. Terima kasih.
